src/chunk-optimization.ts

```diff
@@ -0,0 +1,137 @@
+import Chunk from './Chunk';
+import ExternalModule from './ExternalModule';
+import { OutputOptions } from './rollup';
+
+/*
+ * Given a chunk list, perform optimizations on that chunk list
+ * to reduce the mumber of chunks. Mutates the chunks array.
+ *
+ * Manual chunks (with chunk.chunkAlias already set) are preserved
+ * Entry points are carefully preserved as well
+ *
+ */
+export function optimizeChunks(
+	chunks: Chunk[],
+	options: OutputOptions,
+	CHUNK_GROUPING_SIZE: number
+): Chunk[] {
+	for (let chunkIndex = 0; chunkIndex < chunks.length; chunkIndex++) {
+		const mainChunk = chunks[chunkIndex];
+		const execGroup: Chunk[] = [];
+		mainChunk.postVisitChunkDependencies(dep => {
+			if (dep instanceof Chunk) {
+				execGroup.push(dep);
+			}
+		});
+
+		if (execGroup.length < 2) {
+			continue;
+		}
+
+		let execGroupIndex = 1;
+		let seekingFirstMergeCandidate = true;
+		let lastChunk: Chunk,
+			chunk = execGroup[0],
+			nextChunk = execGroup[1];
+
+		const isMergeCandidate = (chunk: Chunk) => {
+			if (chunk.isEntryModuleFacade || chunk.isManualChunk) {
+				return false;
+			}
+			if (!nextChunk || nextChunk.isEntryModuleFacade) {
+				return false;
+			}
+			if (chunk.getRenderedSourceLength() > CHUNK_GROUPING_SIZE) {
+				return false;
+			}
+			// if (!chunk.isPure()) continue;
+			return true;
+		};
+
+		do {
+			if (seekingFirstMergeCandidate) {
+				if (isMergeCandidate(chunk)) {
+					seekingFirstMergeCandidate = false;
+				}
+				continue;
+			}
+
+			let remainingSize =
+				CHUNK_GROUPING_SIZE - lastChunk.getRenderedSourceLength() - chunk.getRenderedSourceLength();
+			if (remainingSize <= 0) {
+				if (!isMergeCandidate(chunk)) {
+					seekingFirstMergeCandidate = true;
+				}
+				continue;
+			}
+			// if (!chunk.isPure()) continue;
+
+			const chunkDependencies = new Set<Chunk | External>();
+			chunk.postVisitChunkDependencies(dep => chunkDependencies.add(dep));
+
+			const countedChunks = new Set<Chunk | External>([chunk, lastChunk]);
```

Comment:
As I understand these are the "chunks that count", it would be "countingChunks" 😜. No really, how about something like `chunksToBeMerged` as it seems this is exactly what they are? Or, if you really want to stress this is solely about size, it might be `measuredChunks`.

Reply:
They aren't chunks to be merged, as these are chunks in the merged set that are shared dependencies and will not affect the size difference calculation or have already been counted towards the size difference calculation. Chunks not in this set then added towards the combined size check against the chunkGroupingSize limit to determine if the cost of the merge is worthwhile.

I've gone with `ignoreSizeChunks`.

Reply:
Thanks for taking the time coming up with a better name, this really helps a lot!